### ISSUE TITLE
Redirect Loop on Login when active (Shibboleth)

### ISSUE THREAD

**Issue by MadtownLems** [johnbillion/query-monitor#676]
Hi there,

When Query Monitor (3.9) is active on a site with its own domain name in our MultiSite environment, it causes a redirect loop that prevents you from logging in. No PHP errors are logged in error_log, but the browser just shuts down the attempt after a while because it recognizes it's in a loop.  Deactivating QM solves the issue. Reactivating it makes the issue return. 

However, I would be shocked to learn that this isn't related to the fact that we use Shibboleth for SSO, as otherwise it would have been reported by many others.

https://www.shibboleth.net/

I'm sure this is going to be very difficult to debug for that reason. We were most recently using QM 3.8.2 and I don't THINK the issue existed then, but we don't often have QM enabled on sites with custom domains so it's possible we just never noticed.  I'm not sure if you can think of any recent changes that would've caused this? I'm more than willing to hack up a version of QM on our development server if you have suggestions for getting additional debugging information.

Thank you, as always, for your plugin and your help 🙏

**Comment by johnbillion** [comment]
Thanks for the report. If you're seeing this issue on a development server that you can mess around with, can you try removing the collectors (in the `collectors` directory) one by one to narrow down whether one particular collector is causing the problem? You can comment out the collector registration at the bottom of each of the files.

Let me know how you get on.

**Comment by MadtownLems** [comment]
Hiya,

I started commenting out various collectors, and eventually things worked again - but I was never able to fully isolate to a single collector. It seems to be some combination of them.  I'll keep playing and try to isolate it.

In the meantime, I don't know how I missed this in the error logs previously, but the following error is logged when this happens:

[19-May-2022 19:03:36 UTC] PHP Fatal error:  Cannot redeclare wp_get_server_protocol() (previously declared in /wp-includes/load.php:16) in /wp-includes/load.php on line 16

It seems that is typically caused by load.php being included multiple times, but as Query Monitor doesn't directly include load.php, I'm struggling to track that down.

I'll keep working on identifying what combination of collectors needs to be disabled to make logging in work, but thought you might have ideas given the error above in the meantime.

Cheers and thanks!

**Comment by MadtownLems** [comment]
A couple of other observations:

First - I made a few errors/assumptions in the thread title and opening post. My apologies.

1) The first mistake was that this was an infinite redirect loop, but it's not. I believe it's just a fatal PHP error, as mentioned above. The error the browser actually gives me is:

"This page isn't working
foo.bar.cat.edu didn't send any data
ERR_EMPTY_RESPONSE"

2) It's not the actual login process. I can actually log in just fine, as evidenced by me getting all the appropriate WP login cookies. It's actually just viewing a page in the dashboard (once I'm logged in) where things are breaking. (I had assumed it was on login because our SSO instance makes us log into each subdomain independently and until I started hacking up query-monitor, I could never actually see a dashboard screen, but investigating cookies shows me that i'm actually logged in)

Also of note: This is a WordPress Multisite instance, and of course, the issue only exists on sites where QM is active.

Now's where my additional debugging gets even WEIRDER (at least to me).

The problem only exists when I try to view the dashboard with the URL:
foo.bar.cat.edu/wp-admin/

It does NOT happen when I view the same screen with the full foo.bar.cat.edu/wp-admin/index.php (!!)
It does NOT happen when I visit any other page (Add a new post, media library, etc.)

I'll keep poking around, but again, please do let me know if you have any ideas. Thanks again for your time and help. Cheers

**Comment by MadtownLems** [comment]
Well, here's where I'm at...

The error exists when only the following set of Collectors is active: (there may be other combinations as well, but this is the one I stumbled upon.)

admin.php
assets.php
assets_scripts.php
assets_styles.php
db_callers.php
db_queries.php

With the caveat that removing assets.php breaks the script/style collectors, removing any single one of these Collectors will make the error go away and the page load as expected.  But when they are all present, the php error of being unable to redeclare wp_get_server_protocol returns, which, once again, sounds like /wp-includes/load.php is being included a 2nd time.

In an attempt to debug, I hacked up load.php to show the stack trace and die (debug_print_backtrace();
die;) and this is the result:

#0 require() called at [public_html/wp-settings.php:34] 
#1 require_once(public_html/wp-settings.php) called at [wp-config.php:124] 
#2 require_once(wp-config.php) called at [wp-load.php:55] 
#3 require_once(wp-load.php) called at [public_html/wp-admin/admin.php:34] 
#4 require_once(public_html/wp-admin/admin.php) called at [public_html/wp-admin/index.php:10]

I wanted to compare that with the stack trace with the 2nd call to load.php, so used

if ( function_exists( 'wp_get_server_protocol' ) ) {
	debug_print_backtrace();
	die;
}

and it gave me the exact same stacktrace. (I was expecting load.php to be included via a different stack trace, so I was quite surprised.)

I'm really at a loss here. I'm hoping that the list of collectors above that, when active, cause that error, might give you an idea or you can suggest any more debugging steps. As always, cheers and thanks!

**Comment by crstauf** [comment]
The `die()` is probably the issue... trying removing that and see if you get two stack traces.

**Comment by MadtownLems** [comment]
Been working on this for days, and every 'breakthrough' leaves me more and more confused 😅

I also deduced that the error only occurs for us when Query Monitor is used alongside of Really Simple SSL.

I also learned that my attempt to check for the first inclusion of wp-load.php using function_exists( 'wp_get_server_protocol' ) doesn't actually work due to how PHP is parsed, and function_exists will return TRUE if the function is defined later in the file (and not wrapped around its own logic for existing). At least I learned something new!

But, moving on, I have finally been able to get a better idea of what's happening on a given page load when these two plugins are active side by side.  Again, the error only happens when visiting domain.com/wp-admin/ (and adding /index.php to the URL does not have the error).

First...
/wp-admin/index.php
/wp-admin/admin.php
/wp-load.php
/wp-config.php
/wp-settings.php
which requires (not require_once) /wp-includes/load.php

Then, later in the same request...
/index.php
/wp-blog-header.php
/wp-load.php
/wp-config.php
/wp-settings.php
which requires (not require_once) /wp-includes/load.php

As such, load.php is required a second time, causing the fatal error of the first function listed in it not being defined.

When QM is deactivated (but RSSSL remains active), the error goes away. Note that wp-blog-header.php is not used in these situations.
When Really Simple SSL is deactivated (but QM remains active), the error goes away. Note that wp-blog-header.php is not used in these situations either.

But somehow, when both plugins are active (and using the collector configuration from an earlier comment), that 2nd stack trace above (where /index.php calls /wp-blog-header.php, etc) takes place, which results in a second attempt at requiring /wp-includes/load.php.

Please do let me know if that gives you any ideas, things to try, or additional debugging information that would help. As always, cheers and thanks!

**Comment by MadtownLems** [comment]
> Does this happen on any URL other than /wp-admin/ exactly? 
I cannot make the error appear anywhere else. The front-end of the sites are totally fine, including pretty permalinks, taxonomy archives, individual posts, etc.  I've popped all over the dashboard and can write new posts, use the customizer, etc. The only time I can make it happen is /wp-admin/ (and again it's only on the main site or a subsite with its own domain. a site that is in a folder works just fine as foo.bar.cat.edu/subsite/wp-admin/)

> Does the trailing slash affect it?
I'm having a hard time testing this. I commented out the rule that adds a trailing slash to /wp-admin/ from htaccess and tried an incognito window to 'forget' the 301, but it still keeps redirecting there. Perhaps something else is redirecting it?

> I wonder if this is a weird rewrite rule making things go haywire.
This really does feel like that could be the case. Any suggestions on how I can dig a little deeper? These past couple days have really taught me how hard it is to debug without Query Monitor (which is why I'm spending so much time trying to get this resolved!)

**Comment by MadtownLems** [comment]
Just to close the loop here: I spent dozens of hours trying to track this down. Hacking up both plugins. Logging everything I could. I just couldn't make sense of it.

I finally decided to simply add an .htaccess rule that ships anyone from broken URLs to the /index.php equivalent, which seems to be working fine.

If I ever figure out what on earth was going on, I'll be sure to let you know. Otherwise, let's call this closed.

Thanks again for your plugin, your time, and your help!
